osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs

```diff
@@ -367,25 +384,48 @@ private bool performMouseDownActions(MouseButtonEvent e)
         /// <returns>Whether a click selection was active.</returns>
         private bool endClickSelection(MouseButtonEvent e)
         {
-            if (!clickSelectionBegan && !isDraggingBlueprint)
+            // If already handled a selection or drag, we don't want to perform a mouse up / click action.
+            if (clickSelectionHandled || isDraggingBlueprint) return true;
+
+            if (e.Button != MouseButton.Left) return false;
+
+            if (e.ControlPressed)
             {
                 // if a selection didn't occur, we may want to trigger a deselection.
-                if (e.ControlPressed && e.Button == MouseButton.Left)
-                {
-                    // Iterate from the top of the input stack (blueprints closest to the front of the screen first).
-                    // Priority is given to already-selected blueprints.
-                    foreach (SelectionBlueprint<T> blueprint in SelectionBlueprints.AliveChildren.Reverse().OrderByDescending(b => b.IsSelected))
-                    {
-                        if (!blueprint.IsHovered) continue;
-
-                        return clickSelectionBegan = SelectionHandler.MouseUpSelectionRequested(blueprint, e);
-                    }
-                }
+
+                // Iterate from the top of the input stack (blueprints closest to the front of the screen first).
+                // Priority is given to already-selected blueprints.
+                foreach (SelectionBlueprint<T> blueprint in SelectionBlueprints.AliveChildren.Where(b => b.IsHovered).OrderByDescending(b => b.IsSelected))
+                    return clickSelectionHandled = SelectionHandler.MouseUpSelectionRequested(blueprint, e);
 
                 return false;
             }
 
-            return true;
+            if (!wasDragStarted && selectedBlueprintAlreadySelectedOnMouseDown && AllowCyclicSelection)
+            {
+                // If a click occurred and was handled by the currently selected blueprint but didn't result in a drag,
+                // cycle between other blueprints which are also under the cursor.
+
+                // The depth of blueprints is constantly changing (see above where selected blueprints are brought to the front).
+                // For this logic, we want a stable sort order so we can correctly cycle, thus using the blueprintMap instead.
+                IEnumerable<SelectionBlueprint<T>> cyclingSelectionBlueprints = blueprintMap.Values;
+
+                // If there's already a selection, let's start from the blueprint after the selection.
+                cyclingSelectionBlueprints = cyclingSelectionBlueprints.SkipWhile(b => !b.IsSelected).Skip(1);
```

Comment:
I found one more wonky thing: if multiple items are selected, sometimes cycling will happen, and sometimes it will not:

https://user-images.githubusercontent.com/20418176/226414220-10e76db1-f83b-4aa0-9126-ba3fe79d29c5.mp4

It seems that this issue is caused by this one line in particular assuming that the selection consists of exactly one item (the `.Skip(1)` part).

There are two possible fixes. First, the assumption about only one item can be removed by doing this:

```diff
diff --git a/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs b/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs
index 143f343c7d..dece2e0b73 100644
--- a/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs
@@ -411,7 +411,7 @@ private bool endClickSelection(MouseButtonEvent e)
                 IEnumerable<SelectionBlueprint<T>> cyclingSelectionBlueprints = blueprintMap.Values;
 
                 // If there's already a selection, let's start from the blueprint after the selection.
-                cyclingSelectionBlueprints = cyclingSelectionBlueprints.SkipWhile(b => !b.IsSelected).Skip(1);
+                cyclingSelectionBlueprints = cyclingSelectionBlueprints.SkipWhile(b => !b.IsSelected).SkipWhile(b => b.IsSelected);
 
                 // Add the blueprints from before the selection to the end of the enumerable to allow for cyclic selection.
                 cyclingSelectionBlueprints = cyclingSelectionBlueprints.Concat(blueprintMap.Values.TakeWhile(b => !b.IsSelected));
```

which looks like this:

https://user-images.githubusercontent.com/20418176/226414507-f3d56a10-e02f-4a3e-b632-179bcf26cdb3.mp4

Or, cycling can be blocked when multiple items are selected:

```diff
diff --git a/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs b/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs
index 143f343c7d..cb7c083d87 100644
--- a/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs
@@ -401,7 +401,7 @@ private bool endClickSelection(MouseButtonEvent e)
                 return false;
             }
 
-            if (!wasDragStarted && selectedBlueprintAlreadySelectedOnMouseDown && AllowCyclicSelection)
+            if (!wasDragStarted && selectedBlueprintAlreadySelectedOnMouseDown && SelectedItems.Count == 1 && AllowCyclicSelection)
             {
                 // If a click occurred and was handled by the currently selected blueprint but didn't result in a drag,
                 // cycle between other blueprints which are also under the cursor.
```

Which looks like this:

https://user-images.githubusercontent.com/20418176/226414682-2b401f07-762e-4785-b43d-7d0f0e8ef2fd.mp4

Unsure which is preferable; I'm slightly leaning toward the latter.

Reply:
Should be blocked, for sure.